### ISSUE TITLE
[Bug] [Tabs Tray] Undo close tab puts tab at the bottom of the list, rather than in-place

### ISSUE THREAD

**Issue by liuche** [mozilla-mobile/fenix#10986]
* [x] AC bug for investigating https://github.com/mozilla-mobile/android-components/issues/3265

## Steps to reproduce
1. In the new tabs tray, delete a tab
1. Undo deleting the tab from the snackbar

### Expected behavior
Tab restores in-place where it was deleted from

### Actual behavior
Nothing appears to happen, but the tab appears at the bottom of the list, which is confusing (and awkward if you have a lot of tabs)

### Device information
* Android device: ?
* Fenix version: ?

**Comment by darkwing** [comment]
@boek I believe we discovered we can't put the item back in place due to AC.

**Comment by liuche** [comment]
Here's the AC bug for investigating this: https://github.com/mozilla-mobile/android-components/issues/7205

**Comment by softvision-miralobontiu** [comment]
Hi all,
On the latest Beta 5.2.0-beta.2, on Samsung Galaxy Tab S6 (Android 9), after closing tab - Undo, the tab is placed at the top of the list.

**Comment by ekager** [comment]
Looks like the new AC bug for this is https://github.com/mozilla-mobile/android-components/issues/3265

**Comment by madb1lly** [comment]
Hi all,

Maybe the OP needs to be edited to clarify that due to the new tab list layout the old bottom is now the new top.

Cheers 🙂

**Comment by smithi7** [comment]
This is yet another issue that reaĺly boils down to not being able to reorder tabs.

Once we can (again) reorder our tabs, undo after deletion putting restored tab in the wrong place will be merely a brief inconvenience.

I find it hard to believe that a simple hold-and-drag reordering of tabs (or bookmarks, for that matter) is so hard to do when FF did it all those years.

I run with up to 100 tabs, for different projects and interests, and so depend on order to readily find what I'm looking for.  Accidentally deleted tabs - too easy, with no hope of recovery in order, means having to delete then rebuild tabs from collections, which is frustratingly time-consuming.

**Comment by amkcpu** [comment]
Seems like the AC bug is fixed: https://github.com/mozilla-mobile/android-components/issues/3265.

**Comment by JeromeDeBretagne** [comment]
Issue confirmed as still present in:
- Firefox Beta 85 (85.0.0-beta.4) with AC 70.0.5
- Firefox Nightly 210101 with AC 71.0.202101011143133, f31f6d12e

so the above fix in AC is not enough in itself :disappointed:.

@ekager Could this issue be prioritized in a coming sprint :pray: as it leads to a poor experience for users with many tabs?

Hopefully, it should be much smaller than the bigger changes required to [Allow users to manually re-order tabs](https://github.com/mozilla-mobile/fenix/issues/10344) which many of us are eagerly waiting for :heart_eyes:. Thanks.

**Comment by Djfe** [comment]
the thing is: the experience got worse 😅
When the oldest tab was still at the bottom and you restored a tab, it would jump to the restored tab for you.

Since the oldest tab is now at the top (one of the last updates), you have to scroll manually, which takes over a minute for a thousand tabs (I think)

It places the tab at the bottom of the list without telling you where the tab is.
It feels like it didn't got restored at all.

**Comment by mirko77** [comment]
> the thing is: the experience got worse 😅
> When the oldest tab was still at the bottom and you restored a tab, it would jump to the restored tab for you.
> 
> Since the oldest tab is now at the top (one of the last release updates), you have to scroll manually, which takes over a minute for a thousand tabs (I think)
> 
> It places the tab at the bottom of the list without telling you where the tab is.
> It feels like it didn't got restored at all.

Not sure why the tab is not restored at its previous position, it is an array after all. On Gmail, when deleting an email and undoing it, the restored email goes back where it was before. It cannot be that difficult, am I missing something?

**Comment by Amejia481** [comment]
Chating with @jonalmeida he pointed out this functionality is currently blocked on our migration to `BrowserStore`: 
https://github.com/mozilla-mobile/android-components/blob/master/components/feature/session/src/main/java/mozilla/components/feature/session/middleware/undo/UndoMiddleware.kt#L114-L116

**Comment by rocketsroger** [comment]
Waiting on https://github.com/mozilla-mobile/android-components/issues/9593

**Comment by 0x49D1** [comment]
Very annoying thing: tab is moved to the beginning of the list when you accidentally closed it and hit "undo" with no option to reorder it. The user should have some control over the tab list/grid ordering or the "undo" tab option should return the tab to it's previous place in the list/grid.

**Comment by decadent** [comment]
Frankly even if this behaviour was adjusted to put the restored tab at the position where new tabs open, that would be better. (Hopefully that's something like a boolean flag, though it's hard to imagine what code would need to open tabs at the oldest position, at all.)

IIRC I flipped a setting to have new tabs at the bottom (where my fingers are), so it's not a fixed direction. But in any case scrolling back to the old tabs is like visiting a former place of abode with ghosts of a yesteryear and a bygone self.

**Comment by mrBega** [comment]
Hello and sorry if this doesn't belong here. I'm not well versed in the ways of this forum. Moderators, feel free to move this to where it belongs.

I have the problem that when I use the "undo close tab" pop up button, the tab reopenes at the top in the first position. I would like it at the bottom, where new tabs appear, because that is usually the tab I close by mistake. Or even better in the spot where it was when I closed it. I use the grid tab pattern NOT the list one.

Does anyone know where I should post this request? I find this forum hard to navigate on my phone.

And to finish of I would like to extend my deepest gratitude to everyone working on keeping the internet kinda free 💓

**Comment by tukusejssirs** [comment]
> it has been fixed. undo works as expected now in Nightly 93.0a1 (Build #2015832459)

It does not work in my test in 93.0a1 (Build #2015832555) on Android 11.

I’ve opened 4 tabs, closed the third one, in the tab list clicked on the hamburger menu, `Recently closed tabs`, clicked the last closed tab. Result: the tab is reopen as the first (topmost) tab.

**Comment by marcel263** [comment]
> > it has been fixed. undo works as expected now in Nightly 93.0a1 (Build #2015832459)
> 
> It does not work in my test in 93.0a1 (Build #2015832555) on Android 11.
> 
> I’ve opened 4 tabs, closed the third one, in the tab list clicked on the hamburger menu, `Recently closed tabs`, clicked the last closed tab. Result: the tab is reopen as the first (topmost) tab.

that is different from this bug's steps to reproduce, try reproducing by tapping the "undo" snackbar that appears after closing a tab

**Comment by Djfe** [comment]
yep, undo works fine now

and the recently closed tabs feature places them at the beginning of the list (like collections)

I would prefer them to be opened at the end of the list like every new tab but it is indeed a different issue

**Comment by rnc** [comment]
Surely the tab should reopen in the _same_ location (like the desktop version) ; opening in another location, be it the start or the end seems fundamentally broken ?

**Comment by 0x49D1** [comment]
Just saw a few notifications that the bug is fixed. No, it's not: same behavior in Firefox 92. Tested same in Chrome: closed tab opens in the same position in the list as it was before closing; so it's possible, but here in Firefox we have just a frustrating UX :( You just can't order the tabs the way you want for now, that's it.

**Comment by marcel263** [comment]
> Just saw a few notifications that the bug is fixed. No, it's not: same behavior in Firefox 92. Tested same in Chrome: closed tab opens in the same position in the list as it was before closing; so it's possible, but here in Firefox we have just a frustrating UX :( You just can't order the tabs the way you want for now, that's it.

What works now is undo close a tab by tapping "undo" from the snack bar notification. Restoring recently closed from the respective menu is another issue, though clearly related

**Comment by Djfe** [comment]
@0x49D1 the bug was fixed in Nightly (version 93, you are running 92)

**Comment by JeromeDeBretagne** [comment]
Indeed, I can confirm it is working now in Beta 93.0.0-beta.3 (Build #2015834323). I guess this issue can be closed once version 93 gets released as stable @amedyne . It would be great to have a pointer to the actual commit(s) implementing the fix for reference.

**Comment by LaurentiuApahideanSV** [comment]
Verified as fixed on Firefox Preview Nightly 2021-10-04 and Beta 93.0.0-beta.5. When a tab is closed and the undo button is tapped the tab will return to it's previous position.

### Devices used:
- Oppo Find X3 (Android 11)
- OnePlus 6T (Android 9)
